ui/revenuecatui/src/test/kotlin/com/revenuecat/purchases/ui/revenuecatui/PaywallDataValidationTest.kt

```diff
@@ -0,0 +1,138 @@
+
+        val paywall = originalOffering.paywall!!.let { originalPaywall ->
+            val (locale, originalLocalizedConfiguration) = originalPaywall.localizedConfiguration
+            val localizedConfiguration = originalLocalizedConfiguration.copy(
+                features = listOf(
+                    PaywallData.LocalizedConfiguration.Feature(
+                        title = "{{ future_variable }}",
+                        content = "{{ new_variable }}"
+                    ),
+                    PaywallData.LocalizedConfiguration.Feature(
+                        title = "{{ another_one }}"
+                    )
+                )
+            )
+            originalPaywall.copy(localization = mapOf(locale.toString() to localizedConfiguration))
+        }
+
+        val offering = originalOffering.copy(paywall = paywall)
+        val paywallValidationResult = getPaywallValidationResult(offering)
+        verifyPackages(paywallValidationResult.displayablePaywall, originalOffering.paywall!!)
+        compareWithDefaultTemplate(paywallValidationResult.displayablePaywall)
+        assertThat(paywallValidationResult.error).isEqualTo(
+            PaywallValidationError.InvalidVariables(setOf("future_variable", "new_variable", "another_one"))
+        )
+    }
+
+    @Test
+    fun `Unrecognized icons generate default paywall`() {
+        val originalOffering = TestData.offeringWithMultiPackagePaywall
+
+        val paywall = originalOffering.paywall!!.let { originalPaywall ->
+            val (locale, originalLocalizedConfiguration) = originalPaywall.localizedConfiguration
+            val localizedConfiguration = originalLocalizedConfiguration.copy(
+                features = listOf(
+                    PaywallData.LocalizedConfiguration.Feature(
+                        title = "Feature Title",
+                        iconID = "an_unrecognized_icon",
+                    ),
+                )
+            )
+            originalPaywall.copy(localization = mapOf(locale.toString() to localizedConfiguration))
+        }
+
+        val offering = originalOffering.copy(paywall = paywall)
+        val paywallValidationResult = getPaywallValidationResult(offering)
+        verifyPackages(paywallValidationResult.displayablePaywall, originalOffering.paywall!!)
+        compareWithDefaultTemplate(paywallValidationResult.displayablePaywall)
+        assertThat(paywallValidationResult.error).isEqualTo(
+            PaywallValidationError.InvalidIcons(setOf("an_unrecognized_icon"))
+        )
+    }
+
+    private fun getPaywallValidationResult(offering: Offering) = offering.validatedPaywall(
+        currentColorScheme = TestData.Constants.currentColorScheme,
+    )
+
+    private fun verifyPackages(actual: PaywallData, expectation: PaywallData) {
+        assertThat(actual.config.packages).isEqualTo(expectation.config.packages)
+    }
+
+    private fun compareWithDefaultTemplate(displayablePaywall: PaywallData) {
```

Comment:
Nice!! This is a great idea

Reply:
yes... the only issue I am seeing wiht this approach is that it's very hard to see what fails in tests... It will just say that both paywall datas are different and print them both, without printing the actual difference. I might replace it with asserting all properties to be honest